### PR TITLE
cpu/lpc23xx: add MTD wrapper for MCI driver

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

RIOT has a driver for the Memory Card Interface on `lpc23xx`.
It even still works! But it's not integrated into anything.

This adds a MTD wrapper so it can serve as a backend for VFS.

### Testing procedure

Insert a FAT formatted SD card into the SD-Card slot of the `msba2`.

You can run the `examples/filesystem` with

<details><summary>this patch</summary>

```patch
--- a/examples/filesystem/Makefile
+++ b/examples/filesystem/Makefile
@@ -33,9 +33,9 @@ USEMODULE += vfs
 
 # Use a file system
 # USEMODULE += littlefs
-USEMODULE += littlefs2
+# USEMODULE += littlefs2
 # USEMODULE += spiffs
-# USEMODULE += fatfs_vfs
+USEMODULE += fatfs_vfs
 USEMODULE += constfs
 # USEMODULE += devfs
```
</details>

To run `tests/pkg_fatfs_vfs` apply

<details><summary>this patch</summary>

```patch
--- a/tests/pkg_fatfs_vfs/Makefile
+++ b/tests/pkg_fatfs_vfs/Makefile
@@ -19,7 +19,7 @@ ifeq ($(BOARD),native)
   CFLAGS += -DFATFS_IMAGE_FILE_SIZE_MIB=$(FATFS_IMAGE_FILE_SIZE_MIB)
   CFLAGS += -DMTD_SECTOR_NUM=$(MTD_SECTOR_NUM)
 else
-  USEMODULE += mtd_sdcard
+  USEMODULE += mtd
 endif
 ```
</details>

### Issues/PRs references

Sectors / Block numbers might need some work.
It is obvious that the MCI driver was written with fatfs in mind.
`mci_ioctl()` maps directly to `disk_ioctl()`.

However, `fatfs_diskio` then got assumptions about `sdcard_spi` baked in (basically only works with `pages_per_sector = 1`), so don't use this on any SD cards with valuable data.

### CODE REVIEW

File: drivers/mtd_mci/mtd_mci.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    DEBUG("sector count: %lu\n", dev->sector_count);
+    DEBUG("pages / sector: %lu\n", dev->pages_per_sector);
+
+    return 0;
+}
+
+static int mtd_mci_read_page(mtd_dev_t *dev, void *buff, uint32_t page,
+                                uint32_t offset, uint32_t size)
+{
+    (void) dev;
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:57]:
```suggestion
    (void)dev;
```

File: drivers/mtd_mci/mtd_mci.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    DEBUG("sector size: %lu\n", dev->page_size);
+    DEBUG("sector count: %lu\n", dev->sector_count);
+    DEBUG("pages / sector: %lu\n", dev->pages_per_sector);
+
+    return 0;
+}
+
+static int mtd_mci_read_page(mtd_dev_t *dev, void *buff, uint32_t page,
+                                uint32_t offset, uint32_t size)
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:55]:
```suggestion
static int mtd_mci_read_page(mtd_dev_t *dev, void *buff, uint32_t page,
                             uint32_t offset, uint32_t size)
```

File: drivers/mtd_mci/mtd_mci.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    if (mci_read(buff, page, size / SD_HC_BLOCK_SIZE)) {
+        return -EIO;
+    }
+
+    return size;
+}
+
+static int mtd_mci_write_page(mtd_dev_t *dev, const void *buff, uint32_t page,
+                                 uint32_t offset, uint32_t size)
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:73]:
```suggestion
static int mtd_mci_write_page(mtd_dev_t *dev, const void *buff, uint32_t page,
                              uint32_t offset, uint32_t size)
```

File: drivers/mtd_mci/mtd_mci.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+{
+    (void) dev;
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:75]:
```suggestion
    (void)dev;
```

File: drivers/mtd_mci/mtd_mci.c [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    if (mci_write(buff, page, size / SD_HC_BLOCK_SIZE)) {
+        return -EIO;
+    }
+
+    return size;
+}
+
+static int mtd_mci_erase_sector(mtd_dev_t *dev, uint32_t sector, uint32_t count)
+{
+    (void) dev;
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:92]:
```suggestion
    (void)dev;
```

File: drivers/mtd_mci/mtd_mci.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    while (count--) {
+        mci_ioctl(CTRL_ERASE_SECTOR, &sector);
+        ++sector;
+    }
+    return 0;
+}
+
+static int mtd_mci_power(mtd_dev_t *dev, enum mtd_power_state power)
+{
+    (void) dev;
```
Comment (maribu) [inline drivers/mtd_mci/mtd_mci.c:103]:
```suggestion
    (void)dev;
```

**Review by maribu — COMMENTED** [review]
Some comments inline on style.

`examples/filesystem` seems to work as expected. It has been a long time since I last so 8.3 filenames. I'm glad, that this era has passed :-D

**Review by maribu — APPROVED** [review]
ACK

**Comment by maribu** [comment]
I think this could safely be considered a low impact PR, as it only affects MCUs that haven't been sold for quite some time.

**Comment by benpicco** [comment]
I noticed some corruption issues when VFS would attempt to read / write buffers < `SD_HC_BLOCK_SIZE`.
I fixed that by introducing a page buffer so we can always read / write whole pages and then copy a subset of it.

Now read / write works properly:

```
2020-09-25 15:11:40,425 #  mount
2020-09-25 15:11:40,427 # /sda successfully mounted
2020-09-25 15:11:42,394 #  cat /sda/hello.txt
2020-09-25 15:11:42,410 # Hello RIOT!
2020-09-25 15:11:42,411 # This is a simple test.
2020-09-25 15:11:42,411 # Can you read the SD card?
2020-09-25 15:11:42,411 # 
2020-09-25 15:11:42,412 # 
2020-09-25 15:15:36,894 #  vfs w /sda/hello.txt ascii a foobar
2020-09-25 15:15:45,297 #  cat /sda/hello.txt
2020-09-25 15:15:45,297 # Hello RIOT!
2020-09-25 15:15:45,312 # This is a simple test.
2020-09-25 15:15:45,313 # Can you read the SD card?
2020-09-25 15:15:45,313 # 
2020-09-25 15:15:45,313 # 
2020-09-25 15:15:48,041 # foobar
2020-09-25 15:16:06,832 #  vfs w /sda/hello.txt ascii a barfoos
2020-09-25 15:16:09,104 #  cat /sda/hello.txt
2020-09-25 15:16:09,105 # Hello RIOT!
2020-09-25 15:16:09,120 # This is a simple test.
2020-09-25 15:16:09,121 # Can you read the SD card?
2020-09-25 15:16:09,121 # 
2020-09-25 15:16:09,121 # 
2020-09-25 15:16:10,384 # foobarbarfoos
```
(note that `cat` does not print a newline at the end and `vfs w` does not write a newline, so the output only becomes visible with the next command) 

> It has been a long time since I last so 8.3 filenames. I'm glad, that this era has passed :-D

This is actually a config option. `FF_USE_LFN` enables long file names and is disabled by default on RIOT.
The fatfs package is actually quite feature rich. You can even enable exFAT with `FF_FS_EXFAT`.

**Pull request merged**